### DISCUSSION TITLE
What is the rationale for defining a second paramter between Response and Request type?

### DISCUSSION THREAD

**Discussion by darwin403** [axios/axios#5868] [Q&A]
### Describe the issue

The post method has the following type:

```typescript
post<T = any, R = AxiosResponse<T>, D = any>(url: string, data?: D, config?: AxiosRequestConfig<D>): Promise<R>;
```

The infered type ` R = AxiosResponse<T>` must be the 3rd paramter since the `T` (Response Type) and `D` (Request Parameters Type) are sufficient to fix the fully determine.

### Example Code

Currently, if we wish to define the body data type as well, we have to define 3 types. The 2nd parameter should be infered

```typescript
 axios.post<PostResponse, AxiosResponse<PostResponse> , PostBody>("/api/auth/login", { email, password })
      .then(({ data: user }) => console.log(user))
```

### Expected behavior

The ordering of the paramters must change:

```typescript
post<T = any, D = any, R = AxiosResponse<T>>(url: string, data?: D, config?: AxiosRequestConfig<D>): Promise<R>;
```

so that only two types are required to strongly fix the type and ignore the infered type.

```typescript
 axios.post<PostResponse, PostBody>("/api/auth/login", { email, password })
      .then(({ data: user }) => console.log(user))
```

### Additional context/Screenshots

![Screenshot from 2023-04-07 22-42-36](https://user-images.githubusercontent.com/65121170/230649908-6d085aea-d366-48ac-9b97-10a6d03f31c3.png)

**Comment by jasonsaayman** [comment]
The order is historical and changing it would be a breaking TypeScript API change.
`R` is second because Axios lets callers override the full returned promise value, not just `response.data` (`Promise<R>`). Since TypeScript does not support skipping generic parameters, you currently need all three when specifying `D` explicitly.
In many cases, prefer letting `D` infer from the `data` argument, or type the body variable separately:
```ts
const body: PostBody = { email, password };
axios.post<PostResponse>('/api/auth/login', body);
